File: frontend/src/types/attendance.ts
export type AttendanceStatus = 'PRESENT' | 'ABSENT' | 'HALF_DAY' | 'LEAVE'

export const ATTENDANCE_STATUSES: AttendanceStatus[] = ['PRESENT', 'ABSENT', 'HALF_DAY', 'LEAVE']

export interface Attendance {
  id: string
  attendance_date: string
  attendance_taker_id: string
  attendance_taker_name: string
  worker_name: string
  input_parts: string
  total_working_hours: string
  machine_stopped_time: string
  attendance_status: AttendanceStatus
  remarks: string | null
  created_at: string
  updated_at: string
  created_by: string
  updated_by: string | null
}

export interface AttendanceRowInput {
  attendance_date: string
  worker_name: string
  input_parts: number
  total_working_hours: number
  machine_stopped_time: number
  attendance_status: AttendanceStatus
  remarks: string
}

export interface PagedResponse<T> {
  items: T[]
  total: number
  page: number
  page_size: number
}

export interface SupervisorDashboard {
  today_count: number
  present_count: number
  absent_count: number
  total_working_hours: string
  total_machine_stopped_time: string
  total_input_parts: string
  recent_entries: Attendance[]
}

export interface BulkRowError {
  row: number
  worker_name: string
  message: string
  existing_id?: string
}

File: backend/app/schemas/attendance.py
import uuid
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, Field, model_validator

from app.models.attendance import AttendanceStatus


class AttendanceBase(BaseModel):
    attendance_date: date
    worker_name: str = Field(min_length=1, max_length=150)
    input_parts: Decimal = Field(ge=0, default=Decimal("0"), max_digits=10, decimal_places=2)
    total_working_hours: Decimal = Field(ge=0, default=Decimal("0"), max_digits=5, decimal_places=2)
    machine_stopped_time: Decimal = Field(ge=0, default=Decimal("0"), max_digits=5, decimal_places=2)
    attendance_status: AttendanceStatus
    remarks: str | None = Field(default=None, max_length=2000)

    @model_validator(mode="after")
    def apply_business_rules(self) -> "AttendanceBase":
        if self.attendance_status == AttendanceStatus.ABSENT:
            self.input_parts = Decimal("0")
            self.total_working_hours = Decimal("0")
            self.machine_stopped_time = Decimal("0")
        elif self.machine_stopped_time > self.total_working_hours:
            raise ValueError("Machine stopped time cannot exceed total working hours.")
        return self


class AttendanceCreate(AttendanceBase):
    pass


class AttendanceUpdate(AttendanceBase):
    pass


class AttendanceOut(BaseModel):
    id: uuid.UUID
    attendance_date: date
    attendance_taker_id: uuid.UUID
    attendance_taker_name: str
    worker_name: str
    input_parts: Decimal
    total_working_hours: Decimal
    machine_stopped_time: Decimal
    attendance_status: AttendanceStatus
    remarks: str | None
    created_at: datetime
    updated_at: datetime
    created_by: uuid.UUID
    updated_by: uuid.UUID | None


class BulkAttendanceRow(AttendanceBase):
    pass


class BulkAttendanceCreate(BaseModel):
    rows: list[BulkAttendanceRow] = Field(min_length=1, max_length=200)


class BulkRowError(BaseModel):
    row: int
    worker_name: str
    message: str
    existing_id: uuid.UUID | None = None


class VoidAttendanceRequest(BaseModel):
    reason: str = Field(min_length=1, max_length=500)
